File: Unique/Accounts/Events/AuthEvents.cs
using System.Text.Json;
using GTANetworkAPI;
using Unique.Admin.Services;
using Unique.Chat;
using Unique.World;

namespace Unique.Accounts.Events
{
    public class AuthEvents : Script
    {
        private static readonly Vector3 HiddenLoginPos = new Vector3(402.87, -997.81, -99.00);
        private static readonly Vector3 HiddenLoginRot = new Vector3(0, 0, 180f);

        public AuthEvents()
        {
            AccountManager.Load();
        }

        [ServerEvent(Event.PlayerConnected)]
        public void OnPlayerConnected(Player player)
        {
            if (player == null)
                return;

            player.Dimension = 1000u + (uint)player.Id;
            player.Position = HiddenLoginPos;
            player.Rotation = HiddenLoginRot;
            player.Transparency = 0;
            player.SetData("LOGGED_IN", false);
            player.SetData("ACCOUNT_ID", 0);
            player.SetData("ADMIN_LEVEL", 0);
            player.SetData("ADMIN_MODE", false);
            player.SetData("PENDING_SPAWN_SELECTION", false);
        }

        [RemoteEvent("server:auth:ready")]
        public void OnAuthReady(Player player)
        {
            if (player == null)
                return;

            player.TriggerEvent("client:auth:show");
        }

        [ServerEvent(Event.PlayerDisconnected)]
        public void OnPlayerDisconnected(Player player, DisconnectionType type, string reason)
        {
            SavePlayerPosition(player);
        }

        [RemoteEvent("server:auth:register")]
        public void OnRegister(Player player, string firstName, string lastName, string email, string password, string repeatPassword)
        {
            if (player == null)
                return;

            string socialClubId = NAPI.Player.GetPlayerSocialClubId(player).ToString();
            string socialClubName = NAPI.Player.GetPlayerSocialClubName(player)?.Trim();

            firstName = firstName?.Trim();
            lastName = lastName?.Trim();
            email = email?.Trim();
            password = password?.Trim();
            repeatPassword = repeatPassword?.Trim();

            if (!AccountManager.IsValidEmail(email))
            {
                player.TriggerEvent("client:auth:result", false, "E-Mail ist ungueltig.");
                return;
            }

            if (string.IsNullOrWhiteSpace(password) || password.Length < 6)
            {
                player.TriggerEvent("client:auth:result", false, "Passwort muss mindestens 6 Zeichen haben.");
                return;
            }

            if (password != repeatPassword)
            {
                player.TriggerEvent("client:auth:result", false, "Passwoerter stimmen nicht ueberein.");
                return;
            }

            if (AccountManager.EmailExists(email))
            {
                player.TriggerEvent("client:auth:result", false, "Zu dieser E-Mail existiert bereits ein Account.");
                return;
            }

            if (AccountManager.SocialClubExists(socialClubId))
            {
                player.TriggerEvent("client:auth:result", false, "Zu diesem Social Club existiert bereits ein Account.");
                return;
            }

            var account = AccountManager.Create(firstName, lastName, email, password, socialClubName, socialClubId);
            if (account == null)
            {
                player.TriggerEvent("client:auth:result", false, "Account konnte nicht erstellt werden.");
                return;
            }

            BeginCharacterCreation(player, account);
        }

        [RemoteEvent("server:auth:login")]
        public void OnLogin(Player player, string email, string password)
        {
            if (player == null)
                return;

            string socialClubId = NAPI.Player.GetPlayerSocialClubId(player).ToString();
            string socialClubName = NAPI.Player.GetPlayerSocialClubName(player)?.Trim();

            email = email?.Trim();
            password = password?.Trim();

            if (!AccountManager.IsValidEmail(email) || string.IsNullOrWhiteSpace(password))
            {
                player.TriggerEvent("client:auth:result", false, "Bitte gueltige Login-Daten eingeben.");
                return;
            }

            var account = AccountManager.GetByEmail(email);
            if (account == null || !AccountManager.VerifyPassword(account, password))
            {
                player.TriggerEvent("client:auth:result", false, "Falsche E-Mail oder falsches Passwort.");
                return;
            }

            if (account.IsBanned)
            {
                if (IsBanExpired(account))
                {
                    AccountManager.SetBanState(account.AccountId, false, null);
                    account = AccountManager.GetById(account.AccountId);
                }
                else
                {
                    player.SetData("BANNED_SCREEN_ACTIVE", true);
                    player.TriggerEvent("client:auth:banned", BuildBanPayload(account));
                    return;
                }
            }

            if (account == null)
            {
                player.TriggerEvent("client:auth:result", false, "Account wurde nicht gefunden.");
                return;
            }

            if (!AccountManager.IsSocialClubMatch(account, socialClubId))
            {
                player.TriggerEvent("client:auth:result", false, "Dieser Account ist an einen anderen Social Club gebunden.");
                return;
            }

            if (string.IsNullOrWhiteSpace(account.SocialClubId) && !string.IsNullOrWhiteSpace(socialClubId))
                AccountManager.BindSocialClub(account, socialClubName, socialClubId);

            if (!account.CharacterCreated)
            {
                BeginCharacterCreation(player, account);
                return;
            }

            FinishLogin(player, account, false);
        }

        [RemoteEvent("server:character:create")]
        public void OnCreateCharacter(Player player, string characterJson)
        {
            if (player == null || string.IsNullOrWhiteSpace(characterJson))
                return;

            object accountIdValue = NAPI.Data.GetEntityData(player, "ACCOUNT_ID");
            if (!(accountIdValue is int accountId) || accountId <= 0)
            {
                player.TriggerEvent("client:creator:result", false, "Account wurde nicht gefunden.");
                return;
            }

            string firstName;
            string lastName;
            string birthDate;
            string origin;

            try
            {
                using JsonDocument document = JsonDocument.Parse(characterJson);
                JsonElement root = document.RootElement;
                firstName = root.GetProperty("firstname").GetString()?.Trim();
                lastName = root.GetProperty("lastname").GetString()?.Trim();
                birthDate = root.TryGetProperty("birth", out JsonElement birth) ? birth.GetString()?.Trim() : null;
                origin = root.TryGetProperty("origin", out JsonElement originValue) ? originValue.GetString()?.Trim() : null;
            }
            catch
            {
                player.TriggerEvent("client:creator:result", false, "Charakterdaten sind ungueltig.");
                return;
            }

            if (!AccountManager.CompleteCharacter(accountId, firstName, lastName, birthDate, origin, characterJson))
            {
                player.TriggerEvent("client:creator:result", false, "Charakter konnte nicht gespeichert werden.");
                return;
            }

            Account account = AccountManager.GetById(accountId);
            if (account == null)
            {
                player.TriggerEvent("client:creator:result", false, "Account wurde nicht gefunden.");
                return;
            }

            FinishLogin(player, account, true);
        }

        [RemoteEvent("server:spawn:select")]
        public void OnSpawnSelect(Player player, string spawnType)
        {
            if (player == null)
                return;

            object pendingSpawnValue = NAPI.Data.GetEntityData(player, "PENDING_SPAWN_SELECTION");
            if (!(pendingSpawnValue is bool pendingSpawn) || !pendingSpawn)
                return;

            object accountIdValue = NAPI.Data.GetEntityData(player, "ACCOUNT_ID");
            if (!(accountIdValue is int accountId) || accountId <= 0)
                return;

            Account account = AccountManager.GetById(accountId);
            if (account == null)
            {
                player.TriggerEvent("client:spawn:result", false, "Account wurde nicht gefunden.");
                return;
            }

            if (string.Equals(spawnType, "hotel", System.StringComparison.OrdinalIgnoreCase))
            {
                ServerSpawnService.Apply(player);
            }
            else
            {
                player.Dimension = account.Dimension;
                player.Position = new Vector3(account.PosX, account.PosY, account.PosZ);
                player.Rotation = new Vector3(0, 0, account.RotZ);
            }

            player.Health = account.Health <= 0 ? 100 : account.Health;
            player.Armor = account.Armor;
            player.Transparency = 255;
            player.SetData("PENDING_SPAWN_SELECTION", false);

            player.TriggerEvent("client:spawn:hide");
            player.TriggerEvent("client:spawn:resolveGround");
            player.TriggerEvent("client:chat:authState", true);
            player.TriggerEvent("client:hud:authState", true);

            ChatOutput.SendSystem(player, string.Equals(spawnType, "hotel", System.StringComparison.OrdinalIgnoreCase)
                ? "Du bist am Hotel gespawnt."
                : "Du bist an deinem letzten Standort gespawnt.");
        }

        [RemoteEvent("server:auth:banDisconnect")]
        public void OnBanDisconnect(Player player)
        {
            if (player == null)
                return;

            object bannedValue = NAPI.Data.GetEntityData(player, "BANNED_SCREEN_ACTIVE");
            if (bannedValue is bool banned && banned)
                player.Kick("Account gesperrt.");
        }

        private void BeginCharacterCreation(Player player, Account account)
        {
            player.SetData("ACCOUNT_ID", account.AccountId);
            player.SetData("PENDING_CHARACTER_CREATION", true);
            player.Dimension = 2000u + (uint)player.Id;
            ServerSpawnService.Apply(player);
            player.Dimension = 2000u + (uint)player.Id;
            player.Transparency = 255;
            player.Health = 100;
            player.Armor = 0;
            player.TriggerEvent("client:auth:showCreator");
            player.TriggerEvent("client:auth:result", true, "Account erstellt. Erstelle jetzt deinen Charakter.");
        }

        private void FinishLogin(Player player, Account account, bool created)
        {
            player.Name = account.CharacterNameUnderscore;
            player.SetSharedData("DISPLAY_NAME", account.DisplayName);
            player.SetSharedData("ACCOUNT_ID", account.AccountId);
            player.SetSharedData("CHARACTER_NAME", account.CharacterNameUnderscore);
            player.SetSharedData("ADMIN_LEVEL", account.AdminLevel);
            player.SetSharedData("CASH", account.Cash);
            player.SetSharedData("BANK_CASH", account.BankCash);
            player.SetSharedData("HEALTH", account.Health);
            player.SetSharedData("ARMOR", account.Armor);
            player.SetSharedData("SOCIAL_CLUB_ID", account.SocialClubId ?? string.Empty);
            player.SetSharedData("SOCIAL_CLUB_NAME", account.SocialClubName ?? string.Empty);

            player.SetData("LOGGED_IN", true);
            player.SetData("ACCOUNT_ID", account.AccountId);
            player.SetData("ADMIN_LEVEL", account.AdminLevel);
            player.SetData("ADMIN_MODE", false);
            player.SetData("CASH", account.Cash);
            player.SetData("BANK_CASH", account.BankCash);
            player.SetData("PENDING_CHARACTER_CREATION", false);
            player.SetData("PENDING_SPAWN_SELECTION", true);

            player.Dimension = 3000u + (uint)player.Id;
            player.Position = HiddenLoginPos;
            player.Rotation = HiddenLoginRot;
            player.Health = account.Health <= 0 ? 100 : account.Health;
            player.Armor = account.Armor;
            player.Transparency = 0;

            if (!string.IsNullOrWhiteSpace(account.CustomizationJson))
                player.TriggerEvent("client:creator:apply", account.CustomizationJson);

            string socialClubInfo = string.IsNullOrWhiteSpace(account.SocialClubName)
                ? "Kein Social Club gespeichert."
                : $"Social Club: {account.SocialClubName} ({account.SocialClubId})";

            string msg = created
                ? $"Account erstellt. Deine Account-ID ist {account.AccountId}. {socialClubInfo}"
                : $"Erfolgreich eingeloggt. Deine Account-ID ist {account.AccountId}. {socialClubInfo}";

            player.TriggerEvent("client:spawn:show", msg);
            AdminAuthorizationService.RefreshPlayerAdminData(player, account);
        }

        private void SavePlayerPosition(Player player)
        {
            if (player == null)
                return;

            object loggedInValue = NAPI.Data.GetEntityData(player, "LOGGED_IN");
            if (!(loggedInValue is bool loggedIn) || !loggedIn)
                return;

            object pendingSpawnValue = NAPI.Data.GetEntityData(player, "PENDING_SPAWN_SELECTION");
            if (pendingSpawnValue is bool pendingSpawn && pendingSpawn)
                return;

            object adminJailedValue = NAPI.Data.GetEntityData(player, "ADMIN_JAILED");
            if (adminJailedValue is bool adminJailed && adminJailed)
                return;

            object accountIdValue = NAPI.Data.GetEntityData(player, "ACCOUNT_ID");
            if (!(accountIdValue is int accountId) || accountId <= 0)
                return;

            Vector3 position = player.Position;
            float rotZ = player.Rotation.Z;
            int cash = GetIntData(player, "CASH");
            int bankCash = GetIntData(player, "BANK_CASH");
            AccountManager.SavePlayerState(accountId, position.X, position.Y, position.Z, rotZ, player.Dimension, player.Health, player.Armor, cash, bankCash);
        }

        private static int GetIntData(Player player, string key)
        {
            object value = NAPI.Data.GetEntityData(player, key);
            return value is int number ? number : 0;
        }

        private static bool IsBanExpired(Account account)
        {
            if (account == null || string.IsNullOrWhiteSpace(account.BanExpiresAt))
                return false;

            if (!System.DateTime.TryParse(account.BanExpiresAt, null, System.Globalization.DateTimeStyles.RoundtripKind, out System.DateTime expiresAt))
                return false;

            return expiresAt <= System.DateTime.UtcNow;
        }

        private static string BuildBanPayload(Account account)
        {
            if (account == null)
                return "{}";

            string payload =
                "{" +
                $"\"reason\":\"{EscapeJson(account.BanReason ?? "Kein Grund angegeben.")}\"," +
                $"\"banDate\":\"{EscapeJson(account.BanDate ?? System.DateTime.UtcNow.ToString("o"))}\"," +
                $"\"expiresAt\":\"{EscapeJson(account.BanExpiresAt ?? string.Empty)}\"," +
                $"\"admin\":\"{EscapeJson(account.BanAdminName ?? "Unbekannt")}\"," +
                $"\"accountName\":\"{EscapeJson(account.DisplayName)}\"," +
                $"\"accountId\":{account.AccountId}" +
                "}";

            return payload;
        }

        private static string EscapeJson(string value)
        {
            return (value ?? string.Empty)
                .Replace("\\", "\\\\")
                .Replace("\"", "\\\"")
                .Replace("\r", "\\r")
                .Replace("\n", "\\n");
        }
    }
}

File: server/src/features/accounts/auth-events.ts
import { SpawnService } from "../world/spawn-service.js";
import { parseCompleteCharacterDto, parseLoginAccountDto, parseRegisterAccountDto } from "./api/account-dto-parsers.js";
import { validateCompleteCharacterDto, validateLoginAccountDto, validateRegisterAccountDto } from "./account-validators.js";
import { AccountService } from "./account-service.js";
import { HIDDEN_LOGIN_POSITION, emitClient, getSocialClubId, getSocialClubName, getVar, setArmour, setHeading, setVar, vector3 } from "../../runtime/helpers.js";
import type { Account } from "./account.js";
import { FactionService } from "../factions/faction-service.js";
import { PhoneService } from "../phone/phone-service.js";
import { InventoryService } from "../inventory/inventory-service.js";

import { CharacterRepository } from "./character-repository.js";
import type { Character } from "./character.js";

type PlayerMp = any;

type AuthEventDeps = {
  accounts: AccountService;
  characterRepository: CharacterRepository;
  spawns: SpawnService;
  factions: FactionService;
  logError: (message: string, error: unknown) => void;
  systemMessage: (player: PlayerMp, message: string) => void;
  syncFactionMapBlips?: (player: PlayerMp) => Promise<void>;
  phoneService: PhoneService;
  inventory: InventoryService;
};

function buildSpawnPayload(message: string, canUseFactionSpawn: boolean, factionName?: string | null) {
  return JSON.stringify({
    message,
    options: [
      {
        id: "last",
        title: "Letzter Spawn",
        subtitle: "Dein gespeicherter Standort",
        icon: "tab-face",
        disabled: false
      },
      {
        id: "hotel",
        title: "Hotel",
        subtitle: "Server-Spawn",
        icon: "tab-identity",
        disabled: false
      },
      {
        id: "faction",
        title: "Fraktion",
        subtitle: canUseFactionSpawn ? `${factionName ?? "Fraktion"} Spawn` : "Nicht verfuegbar",
        icon: "tab-style",
        disabled: !canUseFactionSpawn
      }
    ]
  });
}

function canUseWardrobeCategory(permissionKeys: string[], category: string) {
  const normalizedCategory = String(category ?? "").trim().toLowerCase();
  if (!permissionKeys.includes("wardrobe_access")) {
    return false;
  }

  if (!normalizedCategory || normalizedCategory === "dienst") {
    return true;
  }

  return permissionKeys.includes(`wardrobe_category_${normalizedCategory}`);
}

async function finishLogin(player: PlayerMp, account: Account, isNew: boolean, deps: AuthEventDeps) {
  setVar(player, "ACCOUNT_ID", account.accountId);
  setVar(player, "EMAIL", account.email);
  setVar(player, "ADMIN_LEVEL", 0);

  const characters = await deps.characterRepository.getByAccountId(account.accountId);
  
  // Explicitly mapping for transparency and type safety
  const charData = (characters || []).map(c => ({
    id: c.characterId,
    firstName: c.firstName,
    lastName: c.lastName,
    cash: Number(c.cash || 0),
    bank: Number(c.bankCash || 0),
    level: Number(c.adminLevel || 1),
    isBanned: !!c.isBanned,
    banReason: c.banReason || null,
    banExpiresAt: c.banExpiresAt || null
  }));

  const charPayload = JSON.stringify(charData || []);
  console.log(`[AUTH-DEBUG] Sending charList (${charData.length} chars) to player ${player.name}`);
  emitClient(player, "client:charselect:show", charPayload);
}

async function finalizeCharacterSpawn(player: PlayerMp, character: Character, deps: AuthEventDeps) {
  const accountId = character.accountId;
  setVar(player, "CHARACTER_ID", character.characterId);
  setVar(player, "LOGGED_IN", true);
  setVar(player, "ADMIN_LEVEL", character.adminLevel);
  setVar(player, "CASH", character.cash);
  setVar(player, "BANK_CASH", character.bankCash);
  
  player.name = `${character.firstName} ${character.lastName}`;
  player.setVariable?.("ACCOUNT_ID", accountId);
  player.setVariable?.("CHARACTER_ID", character.characterId);
  player.setVariable?.("ADMIN_LEVEL", character.adminLevel);
  player.setVariable?.("CASH", character.cash);
  player.setVariable?.("BANK_CASH", character.bankCash);
  player.setVariable?.("LOGGED_IN", true);
  
  await deps.inventory.loadPlayerInventory(player, character.characterId);

  const profile = await deps.factions.getMemberProfile(accountId);
  const permissionKeys = (await deps.factions.getAccountPermissions(accountId)).map((permission) => permission.permissionKey);

  setVar(player, "FACTION_ID", profile?.factionId ?? 0);
  setVar(player, "FACTION_NAME", profile?.factionName ?? "");
  setVar(player, "FACTION_SHORT_NAME", profile?.factionShortName ?? "");
  setVar(player, "FACTION_TYPE", profile?.factionType ?? "");
  setVar(player, "FACTION_RANK", profile?.rankLevel ?? 0);
  setVar(player, "FACTION_RANK_NAME", profile?.rankName ?? "");
  setVar(player, "FACTION_PERMISSIONS", JSON.stringify(permissionKeys));

  player.setVariable?.("FACTION_ID", profile?.factionId ?? 0);
  player.setVariable?.("FACTION_NAME", profile?.factionName ?? "");
  player.setVariable?.("FACTION_RANK", profile?.rankLevel ?? 0);

  player.dimension = 0;
  player.position = HIDDEN_LOGIN_POSITION;
  player.alpha = 0;

  emitClient(player, "client:charselect:hide");
  emitClient(player, "client:spawn:show", buildSpawnPayload("Wo möchtest du starten?", !!profile, profile?.factionName));
  setVar(player, "PENDING_SPAWN_SELECTION", true);
}

// --- UTILITIES KEPT FOR REUSE ---
async function syncFactionState(player: PlayerMp, accountId: number, deps: AuthEventDeps) {
  const profile = await deps.factions.getMemberProfile(accountId);
  const permissions = await deps.factions.getAccountPermissions(accountId);
  const permissionKeys = permissions.map((permission) => permission.permissionKey);

  setVar(player, "FACTION_ID", profile?.factionId ?? 0);
  setVar(player, "FACTION_NAME", profile?.factionName ?? "");
  setVar(player, "FACTION_SHORT_NAME", profile?.factionShortName ?? "");
  setVar(player, "FACTION_TYPE", profile?.factionType ?? "");
  setVar(player, "FACTION_RANK", profile?.rankLevel ?? 0);
  setVar(player, "FACTION_RANK_NAME", profile?.rankName ?? "");
  setVar(player, "FACTION_PERMISSIONS", JSON.stringify(permissionKeys));

  return { profile, permissionKeys };
}

async function syncFactionWardrobeData(player: PlayerMp, accountId: number, permissionKeys: string[], deps: AuthEventDeps) {
  const profile = await deps.factions.getMemberProfile(accountId);
  const wardrobePoints = profile ? await deps.factions.getWardrobePoints(profile.factionId) : [];
  const outfits = profile
    ? (await deps.factions.getOutfits(profile.factionId)).filter((outfit) => canUseWardrobeCategory(permissionKeys, outfit.category))
    : [];
  emitClient(player, "client:factionWardrobe:setData", JSON.stringify({ points: wardrobePoints, outfits }));
}

function isBanExpired(account: Account) {
  if (!account.isBanned || !account.banExpiresAt) {
    return false;
  }
  const expiresAt = Date.parse(account.banExpiresAt);
  return Number.isFinite(expiresAt) && expiresAt <= Date.now();
}

export function registerAuthEvents(deps: AuthEventDeps) {
  mp.events.add("server:auth:ready", (player: PlayerMp) => {
    emitClient(player, "client:auth:show");
  });

  mp.events.add("server:auth:register", (player: PlayerMp, _firstName: string, _lastName: string, email: string, password: string, repeatPassword: string) => {
    void (async () => {
      try {
        const dto = parseRegisterAccountDto(
          { email, password, repeatPassword },
          {
            socialClubId: getSocialClubId(player) || null,
            socialClubName: getSocialClubName(player) || null
          }
        );

        const validation = validateRegisterAccountDto(dto, deps.accounts);
        if (validation.ok === false) {
          emitClient(player, "client:auth:result", false, validation.message);
          return;
        }

        if (await deps.accounts.emailExists(dto.email)) {
          emitClient(player, "client:auth:result", false, "Zu dieser E-Mail existiert bereits ein Account.");
          return;
        }

        if (dto.socialClubId && await deps.accounts.socialClubExists(dto.socialClubId)) {
          emitClient(player, "client:auth:result", false, "Zu diesem Social Club existiert bereits ein Account.");
          return;
        }

        const account = await deps.accounts.create({
          email: dto.email,
          password: dto.password.trim(),
          socialClubName: dto.socialClubName,
          socialClubId: dto.socialClubId
        });

        await finishLogin(player, account, true, deps);
      } catch (error) {
        deps.logError("register failed", error);
        emitClient(player, "client:auth:result", false, "Account konnte nicht erstellt werden.");
      }
    })();
  });

  mp.events.add("server:auth:login", (player: PlayerMp, email: string, password: string) => {
    void (async () => {
      try {
        if (!email || !password) return;
        
        console.log(`[AUTH-DEBUG] Login attempt: ${email.trim()}`);
        const dto = parseLoginAccountDto(
          { email: email.trim(), password },
          {
            socialClubId: getSocialClubId(player) || null,
            socialClubName: getSocialClubName(player) || null
          }
        );

        const validation = validateLoginAccountDto(dto, deps.accounts);
        if (validation.ok === false) {
          emitClient(player, "client:auth:result", false, validation.message);
          return;
        }

        const account = await deps.accounts.getByEmail(dto.email);
        if (!account || !deps.accounts.verifyPassword(account, dto.password)) {
          emitClient(player, "client:auth:result", false, "E-Mail oder Passwort ungültig.");
          return;
        }

        if (account.isBanned) {
          if (isBanExpired(account)) {
            await deps.accounts.setBanState(account.accountId, false, null);
            account.isBanned = false;
          } else {
            setVar(player, "BANNED_SCREEN_ACTIVE", true);
            emitClient(player, "client:auth:banned", JSON.stringify({
               reason: account.banReason,
               banDate: account.banDate,
               expiresAt: account.banExpiresAt
            }));
            return;
          }
        }

        if (!deps.accounts.isSocialClubMatch(account, dto.socialClubId)) {
          emitClient(player, "client:auth:result", false, "Account an anderen Social Club gebunden.");
          return;
        }

        const boundAccount = !account.socialClubId && dto.socialClubId
          ? await deps.accounts.bindSocialClub(account, dto.socialClubName, dto.socialClubId)
          : account;

        console.log(`[AUTH-DEBUG] Login success for ${email.trim()}. Proceeding to charselect.`);
        await finishLogin(player, boundAccount, false, deps);
      } catch (error) {
        deps.logError("login failed", error);
        emitClient(player, "client:auth:result", false, "Login-Dienst vorübergehend nicht erreichbar.");
      }
    })();
  });

  mp.events.add("server:character:create", (player: PlayerMp, characterJson: string) => {
    void (async () => {
      try {
        const accountId = Number(player.getVariable?.("ACCOUNT_ID") ?? 0);
        if (accountId <= 0) {
          emitClient(player, "client:creator:result", false, "Account wurde nicht gefunden.");
          return;
        }

        const dto = parseCompleteCharacterDto(characterJson);
        const validation = validateCompleteCharacterDto(dto, deps.accounts);
        if (validation.ok === false) {
          emitClient(player, "client:creator:result", false, validation.message);
          return;
        }

        const saved = await deps.accounts.completeCharacter(accountId, dto);
        if (!saved) {
          emitClient(player, "client:creator:result", false, "Charakter konnte nicht gespeichert werden.");
          return;
        }

        await finishLogin(player, saved, true, deps);
      } catch (error) {
        deps.logError("character create failed", error);
        emitClient(player, "client:creator:result", false, "Charakterdaten sind ungueltig.");
      }
    })();
  });

  mp.events.add("server:spawn:select", (player: PlayerMp, spawnType: string) => {
    void (async () => {
      try {
        console.log(`[AUTH-DEBUG] Spawn select requested by ${player.name}: ${String(spawnType)}`);
        if (!Boolean(player.getVariable?.("PENDING_SPAWN_SELECTION"))) {
          emitClient(player, "client:spawn:result", false, "Kein Spawn waehlen verfuegbar.");
          return;
        }

        const accountId = Number(player.getVariable?.("ACCOUNT_ID") ?? 0);
        if (accountId <= 0) {
          emitClient(player, "client:spawn:result", false, "Account wurde nicht gefunden.");
          return;
        }

        const account = await deps.accounts.getById(accountId);
        if (!account) {
          emitClient(player, "client:spawn:result", false, "Account wurde nicht gefunden.");
          return;
        }

        const charId = Number(getVar(player, "CHARACTER_ID", 0));
        const character = charId > 0 ? await deps.characterRepository.getById(charId) : null;

        const normalizedSpawn = String(spawnType ?? "").trim().toLowerCase();
        if (normalizedSpawn === "hotel") {
          await deps.spawns.apply(player);
        } else if (normalizedSpawn === "faction") {
          const factionProfile = await deps.factions.getMemberProfile(accountId);
          if (!factionProfile) {
            emitClient(player, "client:spawn:result", false, "Du bist in keiner Fraktion.");
            return;
          }

          const factionSpawn = await deps.factions.getFactionSpawn(factionProfile.factionId);
          if (!factionSpawn) {
            emitClient(player, "client:spawn:result", false, "Deine Fraktion hat keinen Spawnpunkt gesetzt.");
            return;
          }

          player.dimension = factionSpawn.dimension;
          player.position = vector3(factionSpawn.x, factionSpawn.y, factionSpawn.z);
          setHeading(player, factionSpawn.rotZ);
        } else {
          const spawnSrc = character ?? account;
          player.dimension = spawnSrc.dimension;
          player.position = vector3(spawnSrc.posX, spawnSrc.posY, spawnSrc.posZ);
          setHeading(player, spawnSrc.rotZ);
        }

        const healthSrc = character ?? account;
        player.health = healthSrc.health > 0 ? healthSrc.health : 100;
        setArmour(player, healthSrc.armor);
        player.alpha = 255;
        setVar(player, "PENDING_SPAWN_SELECTION", false);

        emitClient(player, "client:spawn:hide");
        emitClient(player, "client:manager:initAll");
        const permissionKeys = (await deps.factions.getAccountPermissions(accountId)).map((permission) => permission.permissionKey);
        await syncFactionWardrobeData(player, accountId, permissionKeys, deps);
        await deps.syncFactionMapBlips?.(player);
        emitClient(player, "client:spawn:resolveGround");
        const customizationJson = character?.customizationJson ?? account.customizationJson;
        if (customizationJson) {
           emitClient(player, "client:creator:apply", customizationJson);
        }
        emitClient(player, "client:chat:authState", true);
        emitClient(player, "client:hud:authState", true);
        deps.systemMessage(
          player,
          normalizedSpawn === "hotel"
            ? "Du bist am Hotel gespawnt."
            : normalizedSpawn === "faction"
              ? "Du bist bei deiner Fraktion gespawnt."
              : "Du bist an deinem letzten Standort gespawnt."
        );

        // --- PHONE INITIALIZATION (POST-SPAWN ONLY) ---
        void (async () => {
          try {
            const phoneNumber = await deps.phoneService.ensurePhoneEntry(accountId, String(player.socialId || accountId));

            setVar(player, "PHONE_NUMBER", phoneNumber);
            setVar(player, "firstName", character?.firstName || account.firstName);
            setVar(player, "lastName", character?.lastName || account.lastName);
            
            deps.logError(`Phone initialized for Account ${accountId} / Character ${charId}`, "INFO");
          } catch (phoneError) {
            deps.logError("Phone initialization failed during spawn", phoneError);
          }
        })();
        // --- END PHONE INITIALIZATION ---
        console.log(`[AUTH-DEBUG] Spawn select completed for ${player.name}: ${normalizedSpawn}`);
      } catch (error) {
        deps.logError("spawn select failed", error);
        emitClient(player, "client:spawn:result", false, "Spawn konnte nicht gesetzt werden.");
      }
    })();
  });

  mp.events.add("server:charselect:select", (player: PlayerMp, characterId: number) => {
    void (async () => {
      try {
        console.log(`[AUTH-DEBUG] Character select requested by ${player.name}: ${Number(characterId)}`);
        const accountId = Number(getVar(player, "ACCOUNT_ID", 0));
        if (accountId <= 0) {
          emitClient(player, "client:charselect:result", false, "Account wurde nicht gefunden.");
          return;
        }

        const char = await deps.characterRepository.getById(characterId);
        if (!char || char.accountId !== accountId) {
          emitClient(player, "client:charselect:result", false, "Charakter nicht gefunden.");
          return;
        }

        if (char.isBanned) {
          emitClient(player, "client:charselect:result", false, `Dieser Charakter ist gebannt. Grund: ${char.banReason}`);
          return;
        }

        await finalizeCharacterSpawn(player, char, deps);
        console.log(`[AUTH-DEBUG] Character select completed for ${player.name}: ${char.characterId}`);
      } catch (error) {
        deps.logError("character select failed", error);
        emitClient(player, "client:charselect:result", false, "Charakterauswahl fehlgeschlagen.");
      }
    })();
  });

  mp.events.add("server:charselect:create", (player: PlayerMp) => {
    const accountId = Number(getVar(player, "ACCOUNT_ID", 0));
    if (accountId <= 0) return;

    emitClient(player, "client:charselect:hide");
    emitClient(player, "client:creator:show");
  });

  mp.events.add("server:auth:banDisconnect", (player: PlayerMp) => {
    if (!Boolean(player.getVariable?.("BANNED_SCREEN_ACTIVE"))) {
      return;
    }

    player.kick?.("Account gesperrt.");
  });
}
